File: src/channel/f_channel_publisher.ts
import { FExecutionContext } from "../execution_context/index.js";

/** Define some kind of Publish-Subscribe pattern. See https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern */
export interface FChannelPublisher<TData> {
	send(executionContext: FExecutionContext, data: TData): Promise<void>;
}

File: freemework_abstractions/src/sync/f_channel_publisher.rs
use futures::future::BoxFuture;

use super::f_channel_event::FChannelEvent;
use super::f_exception::FException;
use super::f_execution_context::FExecutionContext;

///
/// Define some kind of Publish-Subscribe pattern. See https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern
///
pub trait FChannelPublisher<TEventArgs: Send + Sync>: Send + Sync {
    fn publish(
        &self,
        execution_context: FExecutionContext,
        event: FChannelEvent<TEventArgs>,
    ) -> BoxFuture<'_, Result<(), FException>>;
}
